README.md
# Wine-Classification
Classify types of wines according to their chemical characteristics by using machine learning algorithms.


changelog.txt
20170914 - Added headers to data according to description

wine_svm.py
# アルゴリズム: SVM
# 特徴抽出法: PCA / LDA

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# データセット読み込み
dataset = pd.read_csv('wine_data.csv')

# 独立変数と従属変数の分割  
X = dataset.iloc[:, 1:].values
y = dataset.iloc[:, 0].values

# 学習データとテストデータの分割
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

# 標準化
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# 主成分分析(PCA)で次元削減
from sklearn.decomposition import PCA
pca = PCA(n_components = 2) # 可視化するために２次元まで削減する
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)
explained_variance = pca.explained_variance_ratio_

# 線形判別分類(LDA)で次元削減
## from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
## lda = LDA(n_components = 2) # 可視化するために２次元まで削減する
## X_train = lda.fit_transform(X_train, y_train)
## X_test = lda.transform(X_test)

# アルゴリズムに学習させる
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train, y_train)

# テストデータを予測する
y_pred = classifier.predict(X_test)

# 混乱行列を作って精度を測る
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)
# PCAでSVMの結果: 精度 = 97.77%
# LDAでSVMの結果: 精度 = 97.77%

# データの可視化
from matplotlib.colors import ListedColormap
# 学習データ
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green', 'blue')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green', 'blue'))(i), label = j)
## plt.title('SVM with PCA (Training set)')
## plt.xlabel('PC_1')
## plt.ylabel('PC_2')
plt.title('SVM with LDA (Training set)')
plt.xlabel('LD_1')
plt.ylabel('LD_2')
plt.legend()
plt.show()

# テストデータ
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green', 'blue')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green', 'blue'))(i), label = j)
## plt.title('SVM with PCA (Test set)')
## plt.xlabel('PC_1')
## plt.ylabel('PC_2')
plt.title('SVM with LDA (Test set)')
plt.xlabel('LD_1')
plt.ylabel('LD_2')
plt.legend()
plt.show()


wine_logistic_regression.py
# アルゴリズム: ロジスティック回帰 (Logistic Regression)
# 特徴抽出法: PCA / LDA

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# データセット読み込み
dataset = pd.read_csv('wine_data.csv')

# 独立変数と従属変数の分割 
X = dataset.iloc[:, 1:].values
y = dataset.iloc[:, 0].values

# 学習データとテストデータの分割
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

# 標準化
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# 主成分分析(PCA)で次元削減
from sklearn.decomposition import PCA
pca = PCA(n_components = 2) # 可視化するために２次元まで削減する
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)
explained_variance = pca.explained_variance_ratio_

# 線形判別分類(LDA)で次元削減
## from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
## lda = LDA(n_components = 2) # 可視化するために２次元まで削減する
## X_train = lda.fit_transform(X_train, y_train)
## X_test = lda.transform(X_test)

# アルゴリズムに学習させる
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

# テストデータを予測する
y_pred = classifier.predict(X_test)

# 混乱行列を作って精度を測る
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)
# PCAでロジスティクス回帰の結果: 精度 = 97.77%
# LDAでロジスティクス回帰の結果: 精度 = 100.00%

# データの可視化
from matplotlib.colors import ListedColormap
# 学習データ
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green', 'blue')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green', 'blue'))(i), label = j)
plt.title('Logistic Regression with PCA (Training set)')
plt.xlabel('PC_1')
plt.ylabel('PC_2')
## plt.title('Logistic Regression with LDA (Training set)')
## plt.xlabel('LD_1')
## plt.ylabel('LD_2')
plt.legend()
plt.show()

# テストデータ
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green', 'blue')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green', 'blue'))(i), label = j)
plt.title('Logistic Regression with PCA (Test set)')
plt.xlabel('PC_1')
plt.ylabel('PC_2')
## plt.title('Logistic Regression with LDA (Test set)')
## plt.xlabel('LD_1')
## plt.ylabel('LD_2')
plt.legend()
plt.show()


clustering/wine_kmeans_plusplus.py
# アルゴリズム: k-means++ クラスタリング(K-Means++ Clustering)
# 特徴抽出法: PCA

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# データセット読み込み
dataset = pd.read_csv('wine_data.csv')
X = dataset.iloc[:, 1:].values
# クラスタリングは教師なし機械学習なので、従属変数（class）があっても使わないようにします。

# 標準化
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X = sc_X.fit_transform(X)

# 主成分分析(PCA)で次元削減
from sklearn.decomposition import PCA
pca = PCA(n_components = 2) # 可視化するために２次元まで削減する
X = pca.fit_transform(X)
explained_variance = pca.explained_variance_ratio_

# LDAは教師あり機械学習なので、従属変数がない限り利用できません。

# elbow法とWCSS（within-cluster sum of squares）で適切なクラスター数を探す
from sklearn.cluster import KMeans
wcss = []
for i in range(1, 11):
    kmeans = KMeans(n_clusters = i, init = 'k-means++', random_state = 0)
    kmeans.fit(X)
    wcss.append(kmeans.inertia_)
plt.plot(range(1, 11), wcss)
plt.title('The Elbow Method')
plt.xlabel('Number of clusters')
plt.ylabel('WCSS')
plt.show()
# 肘（elbow）のところにWCSSが3なので適切なクラスター数（ｋ）が３クラスターと分かった

# アルゴリズムに学習させる
kmeans = KMeans(n_clusters = 3, init = 'k-means++', random_state = 0)
y_kmeans = kmeans.fit_predict(X)

# クラスターの可視化
plt.scatter(X[y_kmeans == 0, 0], X[y_kmeans == 0, 1], s = 100, c = 'red', label = 'Cluster 1')
plt.scatter(X[y_kmeans == 1, 0], X[y_kmeans == 1, 1], s = 100, c = 'green', label = 'Cluster 2')
plt.scatter(X[y_kmeans == 2, 0], X[y_kmeans == 2, 1], s = 100, c = 'blue', label = 'Cluster 3')
plt.scatter(kmeans.cluster_centers_[:, 0], kmeans.cluster_centers_[:, 1], s = 300, c = 'yellow', label = 'Centroids')
plt.title('Clusters of wine classes')
plt.xlabel('PC_1')
plt.ylabel('PC_2')
plt.legend()
plt.show()

# 参考
# http://rindalog.blogspot.com/2016/08/blog-post_24.html

clustering/wine_hierarchical_clustering.py
# アルゴリズム: 階層的クラスタリング(Hierarchical Clustering)
# 特徴抽出法: PCA

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# データセット読み込み
dataset = pd.read_csv('wine_data.csv')
X = dataset.iloc[:, 1:].values
# クラスタリングは教師なし機械学習なので、従属変数（class）があっても使わないようにします。

# 標準化
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X = sc_X.fit_transform(X)

# 主成分分析(PCA)で次元削減
from sklearn.decomposition import PCA
pca = PCA(n_components = 2) # 可視化するために２次元まで削減する
X = pca.fit_transform(X)
explained_variance = pca.explained_variance_ratio_

# LDAは教師あり機械学習なので、従属変数がない限り利用できません。

# デンドログラムで適切なクラスター数を探す
import scipy.cluster.hierarchy as sch
dendrogram = sch.dendrogram(sch.linkage(X, method = 'ward'))
plt.title('Dendrogram')
plt.xlabel('Classes')
plt.ylabel('Euclidean distances')
plt.show()

# 縦線が高ければ高いほどその間のデータとの関係が遠くなる
# 図中に横線から横線までの一番高い縦線に黒い横線を引いて、３縦線を切ったことで適切なクラスター数が3クラスターと分かった

# アルゴリズムに学習させる
from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters = 3, affinity = 'euclidean', linkage = 'ward')
y_hc = hc.fit_predict(X)

# クラスターの可視化
plt.scatter(X[y_hc == 0, 0], X[y_hc == 0, 1], s = 100, c = 'red', label = 'Cluster 1')
plt.scatter(X[y_hc == 1, 0], X[y_hc == 1, 1], s = 100, c = 'blue', label = 'Cluster 2')
plt.scatter(X[y_hc == 2, 0], X[y_hc == 2, 1], s = 100, c = 'green', label = 'Cluster 3')
plt.title('Clusters of classes')
plt.xlabel('PC_1)')
plt.ylabel('PC_2')
plt.legend()
plt.show()

# もし下の方に横線を引いてみて7クラスターにすればどうなるか

# アルゴリズムに学習させる
from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters = 7, affinity = 'euclidean', linkage = 'ward')
y_hc = hc.fit_predict(X)

# クラスターの可視化
plt.scatter(X[y_hc == 0, 0], X[y_hc == 0, 1], s = 100, c = 'red', label = 'Cluster 1')
plt.scatter(X[y_hc == 1, 0], X[y_hc == 1, 1], s = 100, c = 'blue', label = 'Cluster 2')
plt.scatter(X[y_hc == 2, 0], X[y_hc == 2, 1], s = 100, c = 'green', label = 'Cluster 3')
plt.scatter(X[y_hc == 3, 0], X[y_hc == 3, 1], s = 100, c = 'orange', label = 'Cluster 4')
plt.scatter(X[y_hc == 4, 0], X[y_hc == 4, 1], s = 100, c = 'magenta', label = 'Cluster 5')
plt.scatter(X[y_hc == 5, 0], X[y_hc == 5, 1], s = 100, c = 'pink', label = 'Cluster 6')
plt.scatter(X[y_hc == 6, 0], X[y_hc == 6, 1], s = 100, c = 'black', label = 'Cluster 7')
plt.title('Clusters of classes')
plt.xlabel('PC_1')
plt.ylabel('PC_2')
plt.legend()
plt.show()

# 参考
# https://www.macromill.com/service/data_analysis/d004.html


wine_knn.py
# アルゴリズム: k近傍法 (K-nearest Neighbor)
# 特徴抽出法: PCA / LDA

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# データセット読み込み
dataset = pd.read_csv('wine_data.csv')

# 独立変数と従属変数の分割  
X = dataset.iloc[:, 1:].values
y = dataset.iloc[:, 0].values

# 学習データとテストデータの分割
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

# 標準化
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# 主成分分析(PCA)で次元削減
from sklearn.decomposition import PCA
pca = PCA(n_components = 2) # 可視化するために２次元まで削減する
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)
explained_variance = pca.explained_variance_ratio_

# 線形判別分類(LDA)で次元削減
## from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
## lda = LDA(n_components = 2) # 可視化するために２次元まで削減する
## X_train = lda.fit_transform(X_train, y_train)
## X_test = lda.transform(X_test)

# アルゴリズムに学習させる
from sklearn.neighbors import KNeighborsClassifier
# n_neightbors = numbers of predicting classes (y's classes), metric = 'minkowski' and p = 2 are default parameters
classifier = KNeighborsClassifier(n_neighbors = 3, metric = 'minkowski', p = 2 )
classifier.fit(X_train, y_train)

# テストデータを予測する
y_pred = classifier.predict(X_test)

# 混乱行列を作って精度を測る
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)
# PCAでk-NNの結果: 精度 = 97.77%
# LDAでk-NNの結果: 精度 = 97.77%

# データの可視化
from matplotlib.colors import ListedColormap
# 学習データ
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green', 'blue')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green', 'blue'))(i), label = j)
## plt.title('k-NN with PCA (Training set)')
## plt.xlabel('PC_1')
## plt.ylabel('PC_2')
plt.title('k-NN with LDA (Training set)')
plt.xlabel('LD_1')
plt.ylabel('LD_2')
plt.legend()
plt.show()

# テストデータ
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green', 'blue')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('red', 'green', 'blue'))(i), label = j)
## plt.title('k-NN with PCA (Test set)')
## plt.xlabel('PC_1')
## plt.ylabel('PC_2')
plt.title('k-NN with LDA (Test set)')
plt.xlabel('LD_1')
plt.ylabel('LD_2')
plt.legend()
plt.show()
